### PR TITLE
Implement resource reclamation

### PR DESCRIPTION
**Pull request by omni-viral**

### CODE REVIEW

File: command/src/fence.rs [hunk 1]
```diff
@@ -5,247 +5,141 @@ use {
...
-                    Ok(Err(armed)) => {
-                        std::ptr::write(self, Fence::Armed(armed));
-                        Ok(false)
-                    }
-                    Err((armed, err)) => {
-                        std::ptr::write(self, Fence::Armed(armed));
-                        Err(err)
+    /// Wait for fence to become signaled.
+    /// On success returns complete fence epoch.
+    /// On timeout returns `Ok(None)`
```
Comment (termhn) [inline command/src/fence.rs]: Perhaps we should make our own enum

```rust
pub enum WaitResult {
    Signalled(epoch: FenceEpoch),
    Timeout
}
```

File: command/src/fence.rs [hunk 2]
```diff
@@ -5,247 +5,141 @@ use {
...
+            },
+            _ => panic!("Must be Unsignaled"),
         }
     }
 
-    pub fn armed(self) -> Result<ArmedFence<B>, Self> {
-        match self {
-            Fence::Armed(fence) => Ok(fence),
-            fence => Err(fence),
+    /// Reset signaled fence
```
Comment (termhn) [inline command/src/fence.rs]: Should document panic

File: command/src/fence.rs [hunk 3]
```diff
@@ -5,247 +5,141 @@ use {
...
+            _ => panic!("Must be signaled"),
         }
     }
 
-    pub fn complete(self) -> Result<CompleteFence<B>, Self> {
-        match self {
-            Fence::Complete(fence) => Ok(fence),
-            fence => Err(fence),
+    /// Mark signaled fence as reset.
+    pub unsafe fn mark_reset(&mut self) {
```
Comment (termhn) [inline command/src/fence.rs]: should document panic

File: command/src/fence.rs [hunk 4]
```diff
@@ -5,247 +5,141 @@ use {
...
-                let raw = std::ptr::read(raw);
-                std::ptr::write(self, Fence::Armed(ArmedFence {
-                    raw,
-                    queue,
-                    epoch,
-                }));
-            },
-            _ => panic!("Must be Unarmed"),
+    /// Mark fence as signaled.
+    pub unsafe fn mark_signaled(&mut self) -> FenceEpoch {
```
Comment (termhn) [inline command/src/fence.rs]: Should document panic

File: command/src/fence.rs (hunk 1 above)
Comment (termhn) [reply command/src/fence.rs]: Also should document panic

File: command/src/fence.rs [hunk 5]
```diff
@@ -5,247 +5,141 @@ use {
...
 
     pub fn raw(&self) -> &B::Fence {
-        match self {
-            Fence::Unarmed(fence) => &fence.raw,
-            Fence::Armed(fence) => &fence.raw,
-            Fence::Complete(fence) => &fence.raw,
+        &self.raw
+    }
+    
+    pub fn epoch(&self) -> FenceEpoch {
```
Comment (termhn) [inline command/src/fence.rs]: Document panic

File: command/src/fence.rs [hunk 6]
```diff
@@ -5,247 +5,141 @@ use {
...
+        match self.state {
+            FenceState::Armed(epoch) =>  epoch,
+            _ => panic!("Must be armed"),
         }
     }
-}
\ No newline at end of file
+
+    pub fn into_inner(self) -> B::Fence {
```
Comment (termhn) [inline command/src/fence.rs]: Document panic

File: factory/src/factory.rs [hunk 7]
```diff
@@ -390,86 +407,156 @@ where
...
+                        let family_index = self.families_indices[epoch.queue.family().0];
+                        let lock = epoch_locks[family_index].get_or_insert_with(|| self.epochs[family_index].write());
+                        let queue_epoch = &mut lock[epoch.queue.index()];
+                        *queue_epoch = max(*queue_epoch, epoch.epoch);
+                    }
+                }
+            }
+            WaitFor::All => {
+                for fence in fences {
+                    let epoch = unsafe { /*all fences signaled*/ fence.mark_signaled() };
```
Comment (termhn) [inline factory/src/factory.rs]: We haven't actually checked any of the fences statuses afaict

File: resource/src/resources.rs [hunk 8]
```diff
@@ -245,25 +266,25 @@ where
...
-    /// Image_view can be dropped but this method reduces overhead.
-    pub fn destroy_image_view(
-        &mut self,
-        image_view: image::ImageView<B>,
-    ) {
-        image_view.unescape()
-            .map(|inner| self.dropped_image_views.push(inner));
-    }
+    // /// Destroy image.
+    // /// Image can be dropped but this method reduces overhead.
```
Comment (termhn) [inline resource/src/resources.rs]: do you mean for these to be commented out?

File: factory/src/factory.rs (hunk 7 above)
Comment (omni-viral) [reply factory/src/factory.rs]: Why? We waited for all fences.

File: resource/src/resources.rs (hunk 8 above)
Comment (omni-viral) [reply resource/src/resources.rs]: Yes. Now it is required to know `Epochs`, so those methods have to be rewritten. Gladly no one have been using it.

File: command/src/fence.rs (hunk 1 above)
Comment (omni-viral) [reply command/src/fence.rs]: Too many enums would just require to import more symbols. And this code is mostly for internal use anyway.

**Pull request merged**